src/octtrees.rs

```diff
@@ -0,0 +1,90 @@
+use crate::bbox::BBox;
+use crate::buffers::Buffer;
+use crate::vector::{Vec3, Color};
+
+type Child<T> = Option<Box<T>>;
+
+pub struct OctNode {
+    children: [Child<OctNode>; 8],
+    bounds: BBox,
+    points: Buffer,
+    capacity: usize
+}
+
+impl OctNode {
+    pub fn root_node(radius: f32, capacity: usize) -> Self {
+        Self {
+            children: [None, None, None, None, None, None, None, None],
+            bounds: BBox::new(
+                -radius, radius, 
+                -radius, radius, 
+                -radius, radius),
+            points: Buffer::new(),
+            capacity
+        }
+    }
+
+    pub fn child_node(bounds: BBox, capacity: usize) -> Self {
+        Self {
+            children: [None, None, None, None, None, None, None, None],
+            bounds,
+            points: Buffer::new(),
+            capacity
+        }
+    }
+
+    pub fn is_leaf(&self) -> bool {
+        for child in self.children.iter() {
+            if let Some(_) = child {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    pub fn add(&mut self, point: Vec3, color: Color) {
```

Comment:
I'm not sure how much of this logic should go here and how much should go in the plotters. The goal here is to keep the octree a very basic struct with simple operations, but the orchestration depends on the type of tree.